### PR TITLE
Merge MapLibre Whitepaper into docs

### PR DESCRIPTION
**Pull request by thehoneymad**
I translated the MapLibre whitepaper I wrote and shared on MapLibre Slack to a markdown book. This PR contains the full book as markdown chapters.

I tested it by hosting it locally in my devbox.

This pull request contains the following changesets:
- [MapLibre Whitepaper] Started with a simple introdution and ten thousand foot view section
- [MapLibre Whitepaper] Adding coordiante system section
- [MapLibre Whitepaper] Added expressions and architecture section
- [MapLibre Whitepaper] Added geometry tile worker and android flow chapters
- [MapLibre Whitepaper] Minor updates including file renames and text changes
- [MapLibre Whitepaper] Amended the introduction page

### CODE REVIEW

File: docs/src/SUMMARY.md
Comment (lseelenbinder) [inline docs/src/SUMMARY.md:9]:
```suggestion
    - [Architectural Problems and Recommendations](design/archictural-problems-and-recommendations.md)
```

Also filename probably.

File: docs/src/design/ten-thousand-foot-view.md
Comment (lseelenbinder) [inline docs/src/design/ten-thousand-foot-view.md:35]: Should this also reference tile/source data? (It can't render anything until it has that.)

Generally, I'm not sure if we need to / want to discuss tile sources at this stage. Given it's whole separate system for loading and processing tiles, GeoJSON, etc. from various sources (local or remote), I'm not sure where it should be documented, but I think it probably should be.

File: docs/src/design/ten-thousand-foot-view.md
Comment (lseelenbinder) [inline docs/src/design/ten-thousand-foot-view.md:112]: I'm not sure your definition of SDFs is correct. The PBF contains a glyph _atlas_ (or glyphs). Each element of the atlas is a texture representing the SDF of the character to render.

A SDF uses a field of floats (signed-distance) to represent how a character (or anything else) should be drawn by the GPU. This has the advantage of scaling without artifacts. https://blog.mapbox.com/drawing-text-with-signed-distance-fields-in-mapbox-gl-b0933af6f817 would be the canonical link to explain this.

File: docs/src/design/ten-thousand-foot-view.md
Comment (lseelenbinder) [inline docs/src/design/ten-thousand-foot-view.md:139]: Is it strictly true that it results in `map tiles`? Wouldn't it be true to say _conceptually_ the rendered map has no concept of a tile (other than when you set it in tile mode) rather that is has a viewport on a map (typically composed on a size + center + other camera settings)?

**Review by lseelenbinder — COMMENTED** [review]
Reviewed 10,000 foot view. More to follow as I have time. :smile: 

Thanks for preparing this @thehoneymad!

File: docs/src/design/expressions.md
Comment (OmarShehata) [inline docs/src/design/expressions.md:29]: I would phrase this more as "common uses of expressions are:" because, any property in the mapbox style that takes an expression, you can do something custom/interesting there, beyond the specific use cases mentioned here.

File: docs/src/design/expressions.md
Comment (OmarShehata) [inline docs/src/design/expressions.md:39]: > Expressions use a JSON-like syntax to define

Consider:

> Expressions are defined using a JSON-like syntax.

File: docs/src/design/expressions.md
Comment (OmarShehata) [inline docs/src/design/expressions.md:71]: What is the `object` type?

File: docs/src/design/expressions.md
Comment (OmarShehata) [inline docs/src/design/expressions.md:75]: I know `kind` is the word used in the source code, but a more familiar way to explain this might be to say that you have all these built in functions/operators, such as `coalesce`, `interpolate`, etc.

File: docs/src/design/expressions.md
Comment (OmarShehata) [inline docs/src/design/expressions.md:81]: > an assertion such as `["string", ["get", "feature_property"]]` is necessary

That should be "a coercion" is necessary, right?

File: docs/src/design/expressions.md
Comment (OmarShehata) [inline docs/src/design/expressions.md:94]: Expressions can be used in layout properties too? And in lots of other places (like `filter` etc)

File: docs/src/design/expressions.md
Comment (OmarShehata) [inline docs/src/design/expressions.md:101]: It would be really useful here to include links to the classes in iOS/Android/core/JS. I know the files may move around, but it'd still be incredibly helpful for someone starting out who wants to extend expressions to know what the class names are/where they may be found.

**Review by OmarShehata — COMMENTED** [review]
Hello 👋 , I left a bunch of unsolicited feedback on the expressions doc. Some of it is definitely opinionated so feel free to ignore whatever doesn't make sense here. I've been wanting to contribute docs like this for a while to this project, so thank you @thehoneymad for writing all this up! I learned a lot of cool stuff reading through this.

**Comment by thehoneymad** [comment]
@OmarShehata I deeply appreciate the feedbacks. Please consider them solicited from now on! 🙇 
I will not be able to address all the feedbacks. But I will continuously work on the doument and cut issues to address them. 

I am ❤️ing all the feedbacks I am getting.

**Comment by OmarShehata** [comment]
> But I will continuously work on the doument and cut issues to address them.

That sounds like a great strategy, making incremental improvements! I'd also be happy to try and tackle some of the pages individually once this is merged.

**Comment by thehoneymad** [comment]
> > But I will continuously work on the doument and cut issues to address them.
> 
> That sounds like a great strategy, making incremental improvements! I'd also be happy to try and tackle some of the pages individually once this is merged.

If I could like it twice, I would like it twice.

**Comment by thehoneymad** [comment]
I finally made some time to correct this. :)

File: docs/src/design/ten-thousand-foot-view.md
Comment (thehoneymad) [reply docs/src/design/ten-thousand-foot-view.md:35]: I think it should be documented. I did not dive straight into tiles because I talk about it later when I dive into tile workers. But it is a good concern. I will try to throw a line here or two.

Comment (thehoneymad) [reply docs/src/design/ten-thousand-foot-view.md:35]: I addressed it by adding 2 more lines about Layers and Tiles.

File: docs/src/design/ten-thousand-foot-view.md
Comment (thehoneymad) [reply docs/src/design/ten-thousand-foot-view.md:112]: I oversimplied this one. I deliberately did not talk about the Atlas here. I am reading about that part now so I can document it in a coming PR. 😅

I will read the link you shared and update this section. :)

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:18]: Thank you for the dep read. I fixed this one. :)

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:16]: Noted. The first line did not read easy to me.

> Sources define where the geospatial features to display the map are loaded from.

I wrote the following:

> Layers contains data from a source.  Sources define where the geospatials features displayed on the map are loaded from.

Let me know if this aligns with what you wished for. :)

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:20]: 🙇 Thank you!

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:23]: Yeah. I was dubious about it. I documented it from a perspective of history. DDS was a precursor to expressions. It is there to maintain backward compatibility. It's limitations necessitated the birth of expressions. Expressions are the way to go now. I also do not dive deep into expressions for the same reason.

But there's value in your words. I could not convince myself to pick a definite lane so I kept this section unchanged. Curious to hear what you have to say. 😃

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:71]: The canonical definition of Expressions is rooted in JSON. Like JSON, `object` type  is the mapping key type that maps a set of *keys* to *values*.

I added those lines to the docs too. :)

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:81]: Nice! Yes! 

Thank you again @OmarShehata

**Comment by wipfli** [comment]
@thehoneymad please close the conversations which are resolved from your point of view

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:81]: I also added the following paragraph for future readers:

> You might think this is a *coercion* instead of an *assertion*. If you are 
seeking for coercions, read the upcoming paragraph. Expression names 
that looks like `to-something` are coercions by convention.

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:94]: Ah, I think I went a little too pedantic there. `layer` and `filter` are selctors where `paint` is a property.

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:101]: Oh I wanted to add the classes here honestly. But my knowledge is very limited because I am new to this field. MapLibre Native asks for Graphics, C++, iOS, Android, and Native Interop knowledge all to be have full convergeance of knowledge.

That is why I refrained from adding class names and code pointers. The more I learn, I will update this document with code pointers. I sadly can not act on this feedback now. :)

**Comment by thehoneymad** [comment]
> @thehoneymad please close the conversations which are resolved from your point of view

Yes sir! I am doing that. :) Will upload a new revision soon.

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:14]: Actually, that is very pragmatic. I went a little too theoretical. Your path is very attention grabby. And I LIKE IT.

Comment (thehoneymad) [reply docs/src/design/expressions.md:14]: Allow me to think a moment and incorporate this.

Comment (thehoneymad) [reply docs/src/design/expressions.md:14]: I added your examples in coming revision. :) Thank you for this.

File: docs/src/design/ten-thousand-foot-view.md
Comment (thehoneymad) [reply docs/src/design/ten-thousand-foot-view.md:139]: Yes, the finalized viewport does not have a concept of tiles. I went back and forth with this. Since this is the early text for a reader, I did not dive too deep into rendering implmentation. The code processes one tile at a time and renders the tile in the tile coordinate to produe a single viewport.

But yes, you are right. I do talk about this going forward in the document. :)

Comment (thehoneymad) [reply docs/src/design/ten-thousand-foot-view.md:139]: I am willing to amend the text. I will think more about it. :) Focusing on your SDF comment. I think that one is a bigger crime from my part. 🙈

File: docs/src/design/ten-thousand-foot-view.md
Comment (thehoneymad) [reply docs/src/design/ten-thousand-foot-view.md:112]: Thank you again. I definitely read it wrong. In my head it was about signed distance function. I have another task to add details about text rendering in this document. Probably after I understand the text rendering better.

Comment (thehoneymad) [reply docs/src/design/ten-thousand-foot-view.md:112]: I updated the text to reflect what I learned. Considering this resolved. 😃

**Comment by thehoneymad** [comment]
I addressed all the first series of comments and uploaded new changes. 🙏 
I have more incoming changes on this and other parts of the library. I hope to add dedicated sections for changes from Stamen, and Text Rendering in the future when I understand them better.

If this looks okay to you peeps, I will humbly request a merge to unblock me. :)

File: docs/src/design/expressions.md
Comment (lseelenbinder) [inline docs/src/design/expressions.md:4]: I'm confused what Mapbox Vector Style has to do with Mapbox Vector Tiles? Styles describe how to render any data (including, but not limited to MVTs). Additionally, there are no styles **in** MVTs.

Perhaps you mean something like the following?

```suggestion
vector styles. Mapbox GL Style is used in MapLibre GL renderers.
```

File: docs/src/design/expressions.md
Comment (lseelenbinder) [reply docs/src/design/expressions.md]: The "GeoJSON format" bit is still wrong. GeoJSON is one format, MVT is another, and they have little resemblance to each other (beyond the fact they represent geo data).

File: docs/src/design/geometry-tile-worker.md
Comment (lseelenbinder) [inline docs/src/design/geometry-tile-worker.md:1]: Really helpful overview; this could have saved me so many hours. Thanks!

**Review by lseelenbinder — APPROVED** [review]
Thanks @thehoneymad! I enjoyed reviewing these docs. I think there's probably some remaining deficiencies that will be updated to converge with reality _as our understanding of reality_ converges to the underlying code. 😀

File: docs/src/design/expressions.md
Comment (thehoneymad) [reply docs/src/design/expressions.md:4]: Ah good point. The thing wanted to be explicit about is right now the renderer's definition of Map Style schema is limited to Mapbox Vector Style. There are other style schemas too such as Mapzen.

https://stevebennett.me/2017/08/23/openstreetmap-vector-tiles-mixing-and-matching-engines-schemas-and-styles/

Comment (thehoneymad) [reply docs/src/design/expressions.md:4]: And on the the topic of *understanding of reality*, I whole heartedly agree. I wanted to start somewhere. I would rather be wrong than looking at a vast empty. :D

**Pull request merged**
